### PR TITLE
Add method to recursively remove tabs, and use it for PS8 upgrade

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Theme and module catalog tab links have been removed in 8.0.0, autoupgrade should take that into account
| Type?             | bug fix
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes [#29394](https://github.com/PrestaShop/PrestaShop/issues/29394)
| How to test?      | See how to test in issue [#29394](https://github.com/PrestaShop/PrestaShop/issues/29394)
| Possible impacts? |

### CODE REVIEW

File: upgrade/sql/8.0.0.sql [hunk 1]
```diff
@@ -7,26 +7,18 @@ DROP TABLE IF EXISTS `PREFIX_referrer_shop`;
...
 /* Remove page Referrers */
 ## Remove Tabs
-DELETE FROM `PREFIX_tab` WHERE `class_name` = 'AdminReferrers';
-DELETE FROM `PREFIX_tab_lang` WHERE `id_tab` NOT IN (SELECT id_tab FROM `PREFIX_tab`);
+/* PHP:ps_remove_controller_tab('AdminThemesCatalog'); */;
+/* PHP:ps_remove_controller_tab('AdminModulesCatalog'); */;
+/* PHP:ps_remove_controller_tab('AdminAddonsCatalog'); */;
+/* PHP:ps_remove_controller_tab('AdminReferrers'); */;
 ## Remove Roles
-DELETE FROM `PREFIX_access` WHERE `id_tab` NOT IN (SELECT id_tab FROM `PREFIX_tab`);
```
Comment (matthieu-rolland) [inline upgrade/sql/8.0.0.sql:13]: there is no ´id_tab´ colum in the ps_access table, I'm not sure what this PR was doing... is it a column that existed but was removed from schema?

File: upgrade/php/ps_remove_controller_tab.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        if ($key === 0) {
+            $sqlRoleDeletion .= "slug LIKE '" . $role . "'";
+            continue;
+        }
+        $sqlRoleDeletion .= "OR slug LIKE '" . $role . "'";
+    }
+    Db::getInstance()->execute($sqlRoleDeletion);
+}
+
+function getElementsToBeDeleted($idTab, $idParent, $className, &$tabsToBeDeleted, &$rolesToBeDeleted)
```
Comment (matks) [inline upgrade/php/ps_remove_controller_tab.php:81]: I'm not a fan of passing values by reference but here it does the job 👍

File: upgrade/sql/8.0.0.sql (hunk 1 above)
Comment (matks) [reply upgrade/sql/8.0.0.sql:13]: 🤷‍♀️ 😅

**Review by matks — APPROVED** [review]
OK for me

I wish we could test this

Maybe in another PR we could create a test where
- we run a sqlite database
- we load PrestaShop 8.0.0 database schema
- we run a set of files like this (see below)
- we verify the database schema after process is completed

Example of test script file:
```
/* PHP:ps_remove_controller_tab('AdminThemesCatalog'); */;
/* PHP:do_another_thing('Foo'); */;
/* PHP:drop_your_mic('ps'); */;
/* PHP:ps_remove_controller_tab('AdminReferrers'); */;
```

File: upgrade/php/ps_remove_controller_tab.php (hunk 2 above)
Comment (matthieu-rolland) [reply upgrade/php/ps_remove_controller_tab.php:81]: yes, me neither usually, but it's quite useful in the case of a recursive function, especially when we're not in a class context

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @matthieu-rolland 

I have checked this PR. The issue is well fixed. 
But, i found another issues that i would validate with you the behavior.

Here is my checked list : 

**Upgrade from 1.7.7.8 to 8.0.0** :heavy_check_mark: 

- Check with **php 7.3** : OK 
- Check the upgrade : OK
- Check the issue : OK
- Check with an RTL language (Arabic) for the upgrade : OK
- Check with multistore : OK
- Chefk the BO : OK
- Check the FO : OK
- Check Rollback from 8.0.0 to 1.7.7.8 : OK

https://user-images.githubusercontent.com/92912932/190203009-30fa0942-e468-4256-87ee-4d2320800f9a.mp4



**Upgrade from 1.7.8.7 to 8.0.0** :heavy_check_mark: 

- Check with **php 7.4** : OK 
- Check the upgrade : OK
- Check the issue : OK
- Check with the Spanish language for the upgrade : OK
- Check with multistore : OK
- Chefk the BO : OK
- Check the FO : OK
- Check Rollback from 8.0.0 to 1.7.7.8 : OK

https://user-images.githubusercontent.com/92912932/190203087-f6bee117-b570-4622-a3dd-44c3126de7aa.mp4


**Detected issues :** 

**Issue 1 :** When add a new block on a linklist, there is an exception. It's the same of this fixed [issue](https://github.com/PrestaShop/PrestaShop/issues/28120)  on 5.0.5 released version.

https://user-images.githubusercontent.com/92912932/190201444-a56d9134-280b-4ec3-8456-38550e4d6c6e.mp4

When i want to upgrade the linklist module, the option to upgrade is not displayed.

image

**Without PR** : Another alert when access to linklist page after an upgrade 

https://user-images.githubusercontent.com/92912932/190201579-e4e158dc-6ef0-47c4-85e6-2991090ff6d9.mp4

What do you think in this case ? 


**Issue 2 :** The page is not found : Admin Controller error is displayed in these pages after an upgrade from 1778/1787 to 8.0.0.

**PS** : All problematic modules are uninstalled

**Facebook page : NOK**
image

**Google page : NOK**
image

**Merchant Expertise : NOK**
image

**PrestaShop Metrics : NOK**
image

And, i found this [issue](https://github.com/PrestaShop/PrestaShop/issues/29637) which is not related to this PR.

Could you please check and feedback?

Thank You!

**Comment by matthieu-rolland** [comment]
thank you @HanaRebaiQA !

After investigation, I think we should merge this PR and create issues for your feedbacks, they are mostly not related to this PR. :+1:

**Comment by Robin-Fischer-PS** [comment]
PR is QA :heavy_check_mark: , @matthieu-rolland will give some details about it later.

**Comment by matthieu-rolland** [comment]
About issues found by @HanaRebaiQA :

**Issue 1 with linklist module:** 
The exception is reproduced on dev branch, without the PR, an issue should be created about this.
The impossibility to upgrade should be fixed in the same PR that will fix the impossibility to upgrade, an issue should be created about this too (linklist is a native module, it shouldn't be inactive)

**issue 2**:
We should have a dedicated issue for this, and deal with it in another PR, since this is not a regression from this PR.

Thanx !

**Pull request merged**